src/ParameterLists.f90

```diff
@@ -3062,7 +3129,321 @@ FUNCTION match_ParamType(thisParam,thatParam,prefix) RESULT(bool)
         CLASS DEFAULT
           CONTINUE
       ENDSELECT
-    ENDFUNCTION match_ParamType
+    ENDFUNCTION matchTest_ParamType
+!
+!-------------------------------------------------------------------------------
+!> @brief This function checks the values of thisParam and thatParam and returns 
+!>        if they are equal or approximately equal.
+!> @param thisParam  The parameter list being validated
+!> @param thatParam  The parameter list being checked against
+!> @param bool The logical result of the checked parameters.
+!>
+!> The assumptions of this routine are that the parameters passed in are the 
+!> same extended ParamType.  It also assumes that there is a "gettable" value
+!> that is of thisParam%name on the ParamType.  This function determines
+!> the extended type, then "gets" the appropriate parameter from both lists,
+!> then checks their equivalence.  If they are equal or approximately equal, 
+!> the function results in true.  If not, false.  An error is reported if the 
+!> comparison fails.
```

Comment:
Do we want this to throw an error or is it sufficient to just return false? 

Seems like redundant behavior. If we ALWAYS want this to throw an error on a non-matching condition, then it should not have a return value. If we want it to throw an error sometimes then that is the prerogative of the client code, and that should not be in this routine.

Reply:
I see your point.  One reason I'd argue for keeping both is that this function is called on the parameter level.  If we are comparing a list of parameters, you'd probably like to know all the ones that failed to match on the list at once.  The alternative would be fixing each parameter in serial, with successive reruns of the PL.

 Would it be better to temporarily disable the stop on error for the PL exception handler, and re-enable upon completion of the `verifyList_Paramtype` call?  This more closely matches the `verifyTest_Paramtype`, where all of the parameters in the list are checked with assertions for two given PLs.

Reply:
I would not modify the exception handler because this dissociates the surrogate. (This is the problem with messages getting reported to the log file, or screen or both in a way that is inconsistent).

But sounds like motivation for the error messaging is for debugging only. In that case, you might ifdef or create an optional input arg to modify the exception handler to print all the messages, when a developer wants to do debugging. 

I think the expectation is that `thisParam` and `thatParam` are going to be sane structures, but in testing maybe we want the values to be different. If we are making changes that affect the list structure then we want a strong error so the programmer knows something is up. Can you DBC-ify it? I think that's what we're looking for here.

Reply:
I'll try to explain the functionality better, because I think I'm not doing it well.  Take 2!

The previous `verify` procedure was only to be called for unit tests.  It would compare one parameter list to a reference list and ensure that the structure **and** the values were the same.  At every point, there were assertions to check these values and flag the errors for missing parameters or unequal values.

This procedure was attempted to be used during normal code execution, not a unit test.  It did not work, but the same functionality was needed to compare a parameter structure **and** the values.  If I just needed to compare the structure of a parameter list, I would have used the `validate` procedure.

The desire for having the `verifyList` procedure report _all_ the mismatches is to save the user time updating a single parameter to pass, only to have another fail in the same parameter list.  This behavior is analogous to what we do in XMLProc.  We report all of the errors, and let the user fix all of them, so they do not need to successively rerun cases to fix all of the errors.  The specific case for this is when the input assembly geometry does not match the assembly geometry in the restart file, and there are multiple values that are incorrect.  Does that part make sense?

If we don't want to toggle the `%stoponerror` flag, what would be another good way to report all of the mismatches when calling the `verifyList` procedure?  I'm open to other suggestions if that option screws up the surrogate exception association.

Reply:
Ok, so talked with Brendan about how to handle this, and we came to the conclusion to pass in an exception handler to the `%verifyList` routine.  The problem I ran into when doing this is that when I reassign the original conditions of `eParams` back, the counters all get reset in the unit test, which caused assertions to fail.  There also isn't a way to explicitly set the `%counter` attribute.

The ultimate issue is that there isn't a way to assert that the number of errors matches what was expected in the parameterlist unit test without adding another feature on the exception handler.  The overloaded assignment operation copies the `%counter` array.  We either need a `%setCounter` function, or a special type of assignment function, which I think is unnecessary.  Currently, with my patch to futility and MPACT, I can get the desired behavior in XMLProc, because the `eXML` doesn't stop on errors, and I just change the verbosity on the `eParams` exception handler outside of the calls to `%verifyList`.  The actual feature I need to use in XMLProc is just suppress the parameter list error messages for a call, which I already can do.

Reply:
Updated the code @bkochuna.  Let me know your thoughts.

Reply:
Pinging @bkochuna again.